README.md
# itaro
[![NPM version](https://img.shields.io/npm/v/itaro.svg)](https://npmjs.org/package/itaro)
[![NPM](https://img.shields.io/npm/l/itaro)](./LECENSE)
![David](https://img.shields.io/david/psaren/itaro)
[![npm](https://img.shields.io/npm/dm/itaro)](https://www.npmjs.com/package/itaro)  
taro 一键打包 cli
对于一个由 @tarojs/cli 创建的 taro 项目，可以零配置一键打包多种小程序代码
也可以在taro 项目的 package.json添加 itaro 属性进行个性化配置。

## 安装（install）
``` bash
npm install -g itaro
```

## 命令(commands)
- itaro build

``` bash
$ itaro build
```
- build in watch mode
``` bash
$ itaro watch
```
- show itaro version
``` bash
$ itaro -v
```
- get help
``` bash
$ itaro -h
```
## 配置(config)
在 taro 项目的 package.json，添加配置项
``` javascript
// package.json
{
  ...
  "itaro": {
    "build": {
      "output": "itaro", // 打包默认的输出根目录
      "excludes": ["rn"], // 不需要打包的小程序名称
      "options": ["alipay", "tt", "qq", "weapp", "rn"] // 需要打包的小程序名称
    }
  }
}
```
`options` 除了可以是 `weapp`, `alipay` 等字符串也可以是
``` javascript
"options": [
  {
    // 构建名称
    "name": "weapp",
    // 构建命令(默认为 npm run build:构建名称)
    "command": "npm run build:weapp", 
     // 输出目录(默认为 输出根目录/构建名称)
    "output": "itaro/weapp",
  }
]
```
## 实现原理
通过修改 `config/index.js outputRoot` 属性来实现
![itaro-build](src/imgs/itaro-build.png)

``` javascript
{
  "output": "itaro",
  "options": ["weapp", "alipay", "swan", ...]
}

runTask("weapp")
// 修改 outputRoot: "itaro/weapp",
// 执行构建命令 npm run build:weapp
// 还原 outputRoot: "dist",
runTask("alipay")
// 修改 outputRoot: "itaro/alipay",
// 执行构建命令 npm run build:alipay
// 还原 outputRoot: "dist",
runTask("swan")
// 修改 outputRoot: "itaro/swan",
// 执行构建命令 npm run build:swan
// 还原 outputRoot: "dist",

// ...
```
## CHANGELOG
[CHANGELOG](CHANGELOG.md)
## LECENSE
[MIT](LECENSE)

CHANGELOG.md
## 0.1.1 (2020-06-03)


### Features

* 修改默认输出目录配置 ([2beae51](https://github.com/psaren/itaro/commit/2beae51c4970928f198bd8b57af0e81ca5013083))
* **build:** 调整任务执行间隔 ([f4a9b1a](https://github.com/psaren/itaro/commit/f4a9b1a4f0a19c92dacd5444095932032e762b8a))
* **watch:** watch .vue file ([989346d](https://github.com/psaren/itaro/commit/989346d2659dc3ede83ed3d7444b0a21c46baa7f))



# 0.1.0 (2020-06-02)


### Features

* **watch:** watch .vue file ([989346d](https://github.com/psaren/itaro/commit/989346d2659dc3ede83ed3d7444b0a21c46baa7f))
* 修改默认输出目录配置 ([2beae51](https://github.com/psaren/itaro/commit/2beae51c4970928f198bd8b57af0e81ca5013083))





config/index.js
const config = {
  projectName: 'taro-project-v2.2.6',
  date: '2020-6-1',
  designWidth: 750,
  deviceRatio: {
    '640': 2.34 / 2,
    '750': 1,
    '828': 1.81 / 2
  },
  sourceRoot: 'src',
  outputRoot: ./output/weapp,
  babel: {
    sourceMap: true,
    presets: [
      ['env', {
        modules: false
      }]
    ],
    plugins: [
      'transform-decorators-legacy',
      'transform-class-properties',
      'transform-object-rest-spread',
      ['transform-runtime', {
        'helpers': false,
        'polyfill': false,
        'regenerator': true,
        'moduleName': 'babel-runtime'
      }]
    ]
  },
  plugins: [],
  defineConstants: {
  },
  mini: {
    postcss: {
      pxtransform: {
        enable: true,
        config: {}
      },
      url: {
        enable: true,
        config: {
          limit: 10240 // 设定转换尺寸上限
        }
      },
      cssModules: {
        enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
        config: {
          namingPattern: 'module', // 转换模式，取值为 global/module
          generateScopedName: '[name]__[local]___[hash:base64:5]'
        }
      }
    }
  },
  h5: {
    publicPath: '/',
    staticDirectory: 'static',
    postcss: {
      autoprefixer: {
        enable: true,
        config: {
          browsers: [
            'last 3 versions',
            'Android >= 4.1',
            'ios >= 8'
          ]
        }
      },
      cssModules: {
        enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
        config: {
          namingPattern: 'module', // 转换模式，取值为 global/module
          generateScopedName: '[name]__[local]___[hash:base64:5]'
        }
      }
    }
  }
}

module.exports = function (merge) {
  if (process.env.NODE_ENV === 'development') {
    return merge({}, config, require('./dev'))
  }
  return merge({}, config, require('./prod'))
}


src/commands/build.bak.ts
import * as path from 'path'
import * as fs from 'fs'
import * as shell from 'shelljs'
import * as chalk from 'chalk'
import { isObject, isTest, sleep } from '../utils'
const rootCwd = process.cwd()
const pkgJson = require(path.resolve(rootCwd, './package.json'))
const numCPUs = require('os').cpus().length
let runningTasks = 0 // 正在运行的任务数量
const tasks:any[] = [] // 任务列表
const restoreTimeout = 2000 // 还原配置的时间

const VALID_OPTIONS = [
  'weapp',
  'swan',
  'alipay',
  'tt',
  'h5',
  'rn',
  'qq',
  'quickapp',
  'jd'
]

interface Item {
  name: string;
  command: string;
  output: string;
}

function getBuildOptions (arr: (Item|string)[], output): Item[] {
  if (!Array.isArray(arr)) return []
  const res: Item[] = []
  arr.forEach(item => {
    if (typeof item === 'string') {
      res.push({
        name: item,
        command: `npm run build:${item}`,
        output: `${output}/${item}`
      })
    } else if (isObject(item)) {
      if (item.name) {
        const resultItem: Item = {
          command: `npm run build:${item.name}`,
          output: `${output}/${item.name}`,
          ...item
        }
        res.push(resultItem)
      }
    }
  })
  return res
}

function getBuildConfig () {
  let buildConfig = pkgJson.taro && isObject(pkgJson.taro.build) ? pkgJson.taro.build : {}

  const defaultConfig = {
    output: './output',
    excludes: [],
    options: VALID_OPTIONS
  }

  buildConfig = Object.assign({}, defaultConfig, buildConfig)
  return buildConfig
}

// 生成新的 build 任务
function genNewTask (name: string, command: string, output: string) {
  return async () => {
    try {
      const outputPath = path.resolve(rootCwd, output)
      // 检查 npm 命令是否存在
      if (pkgJson.scripts[command.replace('npm run ', '')]) {
        const configPath = path.resolve(rootCwd, './config/index.js')
        // const configPath = path.resolve(__dirname, '../../config/index.js')
        // 读取 config
        const config = fs.readFileSync(configPath, { encoding: 'utf8' })
        // 修改 config.outputRoot
        const newConfig = config.replace(/outputRoot:\s('|")(\w+\/?)+('|"),/gm, `outputRoot: '${output}',`)
        // 更新 config
        fs.writeFileSync(configPath, newConfig, { encoding: 'utf8' })

        shell.exec(command, { async: true }, () => {
          runningTasks--
          console.log(chalk.green(`build ${name} succefully!`))
          runNextTasks()
        })
        // 创建项目产出目录
        shell.mkdir('-p', outputPath)
        setTimeout(() => {
          // 还原 config
          fs.writeFileSync(configPath, config, { encoding: 'utf8' })
        }, restoreTimeout)
      } else {
        runningTasks--
        isTest && console.log(chalk.red('缺少scripts >> '), command)
      }
    } catch (err) {
      runningTasks--
      console.log(chalk.red(`build ${name} failure!`))
      console.error(err)
    }
  }
}

// 执行下一个 build 任务
async function runNextTasks () {
  while (tasks.length > 0 && runningTasks < numCPUs) {
    runningTasks++
    const task = tasks.shift()
    await sleep(restoreTimeout + 1000)
    task()
  }
}

export default function build () {
  const config = getBuildConfig()

  shell.exec('chcp 65001')

  const options = getBuildOptions(config.options, config.output)

  for (const option of options) {
    const { name, command, output } = option
    if (config.excludes.includes(name) || !VALID_OPTIONS.includes(name)) continue
    tasks.push(genNewTask(name, command, output))
  }
  runNextTasks()
}


src/commands/build.ts
import * as path from 'path'
import * as fs from 'fs'
import * as shell from 'shelljs'
import * as chalk from 'chalk'
import { isObject, isTest, sleep, isArray } from '../utils'
const rootCwd = process.cwd()
const pkgJson = require(path.resolve(rootCwd, './package.json'))
const numCPUs = require('os').cpus().length
const itaroBuildTime = 'Itaro Build Time'
const defaultOutput = 'itaro'

const VALID_OPTIONS = [
  'weapp',
  'swan',
  'alipay',
  'tt',
  'h5',
  'rn',
  'qq',
  'quickapp',
  'jd'
]

interface BuilderInterFace {
  runningTasks: number;
  tasks: any[];
  restoreTimeout: number;
  remainingTasks: number;
  isWatch: boolean;
}

interface Item {
  name: string;
  command: string;
  output: string;
}

export default class Builder implements BuilderInterFace {
  runningTasks: number;
  tasks: any[];
  restoreTimeout: number;
  remainingTasks: number;
  isWatch: boolean;
  constructor(isWatch = false) {
    // 正在运行的任务数量
    this.runningTasks = 0
    // 任务列表
    this.tasks = []
    // 还原配置的时间
    this.restoreTimeout = 3000
    // 是否开启监听
    this.isWatch = isWatch
    // 剩余任务数量，会在 run 方法执行第一个任务前赋值
    this.remainingTasks = 0
  }

  getBuildOptions(arr: (Item | string)[], output): Item[] {
    if (!isArray(arr)) return []
    const res: Item[] = []
    arr.forEach(item => {
      if (typeof item === 'string') {
        res.push({
          name: item,
          command: `npm run build:${item}`,
          output: `${output}/${item}`
        })
      } else if (isObject(item)) {
        if (item.name) {
          const resultItem: Item = {
            command: `npm run build:${item.name}`,
            output: `${output}/${item.name}`,
            ...item
          }
          res.push(resultItem)
        }
      }
    })
    return res
  }

  getBuildConfig() {
    let buildConfig = pkgJson[defaultOutput] && isObject(pkgJson[defaultOutput].build) ? pkgJson[defaultOutput].build : {}

    const defaultConfig = {
      output: defaultOutput,
      excludes: [],
      options: VALID_OPTIONS
    }

    buildConfig = Object.assign({}, defaultConfig, buildConfig)
    return buildConfig
  }

  // 生成新的 build 任务
  genNewTask(name: string, command: string, output: string) {
    return () => new Promise((resolve, reject) => {
      try {
        const outputPath = path.resolve(rootCwd, output)
        // 检查 npm 命令是否存在
        if (pkgJson.scripts[command.replace('npm run ', '')]) {
          const configPath = path.resolve(rootCwd, './config/index.js')
          // 读取 config
          const config = fs.readFileSync(configPath, { encoding: 'utf8' })
          // 修改 config.outputRoot
          const newConfig = config.replace(/outputRoot:\s('|")(\w+\/?)+('|"),/gm, `outputRoot: '${output}',`)
          // 更新 config
          fs.writeFileSync(configPath, newConfig, { encoding: 'utf8' })

          shell.exec(command, { async: true }, () => {
            this.reduceTasksNum()
            console.log(chalk.green(`build ${name} succefully!`))
            resolve(true)
            this.runNextTasks()
          })
          // 创建项目产出目录
          shell.mkdir('-p', outputPath)
          setTimeout(() => {
            // 还原 config
            fs.writeFileSync(configPath, config, { encoding: 'utf8' })
          }, this.restoreTimeout)
        } else {
          resolve(false)
          this.reduceTasksNum()
          isTest && console.log(chalk.red('缺少scripts >> '), command)
        }
      } catch (err) {
        resolve(err)
        this.reduceTasksNum()
        console.log(chalk.red(`build ${name} failure!`))
        console.error(err)
      }
    })
  }

  reduceTasksNum() {
    this.remainingTasks--
    if (this.runningTasks > 0) {
      this.runningTasks--
    }
  }

  // 执行下一个 build 任务
  async runNextTasks() {
    const { tasks, runningTasks, restoreTimeout } = this
    while (tasks.length > 0 && runningTasks < numCPUs) {
      this.runningTasks++
      const task = this.tasks.shift()
      await sleep(restoreTimeout + 1000)
      task().then(() => {
        if (this.remainingTasks < 1) {
          console.log('')
          console.timeEnd(itaroBuildTime)
          if (this.isWatch) {
            console.log('')
            console.log(chalk.green('itaro is watching files change...'))
          }
        }
      })
    }
  }

  async run() {
    console.log(chalk.green('itaro build running...'))
    console.time(itaroBuildTime)
    const config = this.getBuildConfig()

    shell.exec('chcp 65001', { silent: true })

    const options = this.getBuildOptions(config.options, config.output)
    for (const option of options) {
      const { name, command, output } = option
      if (config.excludes.includes(name) || !VALID_OPTIONS.includes(name)) continue
      this.tasks.push(this.genNewTask(name, command, output))
    }
    this.remainingTasks = this.tasks.length
    this.runNextTasks()
  }
}
